language-plutus-core/src/Language/PlutusCore/Check/ValueRestriction.hs

```diff
@@ -0,0 +1,34 @@
+-- | The value restriction check.
+
+{-# LANGUAGE LambdaCase #-}
+
+module Language.PlutusCore.Check.ValueRestriction
+    ( checkValueRestriction
+    , checkValueRestrictionProgram
+    ) where
+
+import           Language.PlutusCore.Error
+import           Language.PlutusCore.Type
+import           Language.PlutusCore.View  (isTermValue)
+
+import           Control.Monad.Error.Lens
+import           Control.Monad.Except
+import           Data.Foldable             (traverse_)
+import           Data.Functor.Foldable
+
+-- | Check whether a term satisfies the value restriction.
+checkValueRestriction
+    :: (AsValueRestrictionError e tyname ann, MonadError e m) => Term tyname name ann -> m ()
+checkValueRestriction = para $ \case
+    TyAbsF ann name _ (t, r) -> do
+        case t of
+            TyAbs{}           -> pure ()
+            _ | isTermValue t -> pure ()
+            _                 -> throwing _ValueRestrictionError $ ValueRestrictionViolation ann name
+        r
```

Comment:
Just changed that. Do not merge, I'll write a comment tomorrow on what's going on here.